tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Refiner.scala

```diff
@@ -69,10 +69,12 @@ final class Refiner(semantics: Semantics, outputMode: OutputMode) {
     }
 
     val kind = {
-      if (classDef.kind == ClassKind.ModuleClass && !info.isModuleAccessed)
-        ClassKind.Class
-      else
+      if (classDef.kind.hasModuleAccessor && !info.isModuleAccessed) {
+        if (classDef.kind == ClassKind.ModuleClass) ClassKind.Class
+        else ClassKind.JSClass
+      } else {
         classDef.kind
+      }
```

Comment:
Have you considered unifying this code with the one in the linker? It seems like a good candidate for a helper.